File: app/src/main/java/com/romellfudi/callbacks/MainActivity.java
package com.romellfudi.callbacks;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import java.text.MessageFormat;
import java.text.ParseException;

public class MainActivity extends AppCompatActivity {
    TextView txt;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        txt = (TextView) findViewById(R.id.txt);
    }

    public void Callback(View view) {
        myController.getInstance().guardarData(txt.getText().toString(),
                new Callback() {
                    @Override
                    public void successed(Object... objects) {
                        show((String) objects[0]);
                        txt.append(objects[1].toString());
                    }

                    @Override
                    public void failed(Object... objects) {
                        show((String) objects[0]);

                    }
                });
    }

    public void EnviarCallback(View view) {

        //sendCallback
        myController.getInstance().enviarData(2, new float[]{0, 8}, new EnviarCallback() {
            @Override
            public void errorAlParsear(Exception exception) {
                if (exception != null)
                    show(exception.getMessage());

            }

            @Override
            public void successed(Object... objects) {
                show(objects[0].toString());

            }

            @Override
            public void failed(Object... objects) {
                show(objects[0].toString());
            }
        });
    }

    public void RestCallBack(View view) {
        myController.getInstance().enviarData(0, new float[]{})
                .enqueRest(new RestCallBack() {
                    @Override
                    public void httpOK(Object... objects) {
                        int num = (int) objects[0];
                        float[] floats = (float[]) objects[1];
                        Model model = (Model) objects[2];

                    }

                    @Override
                    public void errorHttp(Exception exception, Object... objects) {
                        if (exception == null)
                            show(exception.getMessage());
                    }

                    @Override
                    public void successed(Object... objects) {
                        show(objects[0].toString());
                    }

                    @Override
                    public void failed(Object... objects) {
                        show(objects[0].toString());
                    }
                });

    }

    public void MathCallback(View view) {
        //MathCallback
        myController.getInstance().resolverProblema(3, 4, new MathCallback() {
            @Override
            void Mathsuccessed(Object... objects) {
                show(objects[0].toString());
            }

            @Override
            void Mathfailed(NumberFormatException exception, Object... objects) {
                show(exception.getMessage());
            }
        });
    }

    public void ObjectCallback(View view) {
//        ObjectCallback
        myController.getInstance().resolverObject(null, new ObjectCallback() {
            @Override
            public void successed(Object... objects) {
                show(objects[0].toString());
            }
        });
    }

    public void AbstractCallback(View view) {
//        AbstractCallback
        myController.getInstance().resolverAbstract(3, "34", new AbstractCallback() {
            @Override
            public void successed(Object... objects) {
                show(objects[0].toString());
            }

            @Override
            public void failed(Exception exception, Object... objects) {
                if (exception!=null)
                    show(exception.getMessage());
            }
        });

    }

    public void MethodParse(View view){
        myController.getInstance().getParsedObject("", new myController.MethodCall(){
            @Override
            public void onFail(Object... objects) {
                objects[0].toString();
            }

            @Override
            public void onParse(ParseException exception, Object... objects) {
                show(exception.getMessage());
            }

            @Override
            public void onSuccess(Object... objects) {
                objects[0].toString();
            }
        });
    }

    public void OtherMethodParse(View view){
        myController.getInstance().getHttp("", new myController.HttpCall<Model>() {
            @Override
            public void errorHTTP(String codeError, Exception exception, Object... objects) {
                show(codeError+" - "+exception.getMessage()+" - "+objects[0].toString());
            }

            @Override
            public void onResponse(Model model, Object... objects) {
                show(objects[0].toString());
            }
        });
    }

    public void twoSteps(View view){
        myController.getInstance().twoSteps(new PreCallback() {
            @Override
            public void successed(Object... objects) {
                show(objects[0].toString());
            }

            @Override
            public void failed(Object... objects) {
                show(objects[0].toString());
            }
        }, new PosCallback() {
            @Override
            public void successed(Object... objects) {
                show(objects[0].toString());
            }

            @Override
            public void failed(Object... objects) {
                show(objects[0].toString());
            }
        });
    }

    public void interfaces(View view){
        new TemplateController<Model,Model2>()
                .methodStatickCall(new TemplateController.TemplateCallback<Model, Model2>() {
                    @Override
                    public Model2 methodPoorReturnY(Model model) {
                        return new Model2();
                    }

                    @Override
                    public Model methodSuperReturnA(Model2 model2) {
                        return new Model();
                    }
                });
    }

    public void returnCallback(View view) {
        final int value =3;
        myController.getInstance().returnTripleCallback(value,new OtherCallback(){

            @Override
            public void returns(int it) {
                show(MessageFormat.format("The triple of {0} is {1}", value, it));
            }
        });
    }

    private void show(String string) {
        Toast.makeText(MainActivity.this, string, Toast.LENGTH_SHORT).show();
    }
}

File: app-kotlin/src/main/java/com/romellfudi/callbacks/MainActivity.kt
package com.romellfudi.callbacks

import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.widget.Toast
import com.romellfudi.callbacks.MyController.HttpCall
import com.romellfudi.callbacks.MyController.MethodCall
import com.romellfudi.callbacks.databinding.ActivityMainBinding
import kotlinx.android.synthetic.main.activity_main.*
import java.text.ParseException

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (DataBindingUtil.setContentView(this, R.layout.activity_main)
                as ActivityMainBinding).also { it.main = this }
    }

    fun Callback() {
        MyController.guardarData(txt.text.toString(),
                object : Callback {
                    override fun successed(vararg objects: Any) {
                        show(objects[0] as String)
                        txt.append(objects[1].toString())
                    }

                    override fun failed(vararg objects: Any) {
                        show(objects[0] as String)
                    }
                })
    }

    fun EnviarCallback() {
        MyController.enviarData(2, floatArrayOf(0f, 8f), object : EnviarCallback {
            override fun errorAlParsear(exception: Exception?) {
                show(exception?.message ?: "No exception found")
            }

            override fun successed(vararg objects: Any) {
                show(objects[0].toString())
            }

            override fun failed(vararg objects: Any) {
                show(objects[0].toString())
            }
        })
    }

    fun RestCallBack() {
        MyController.apply {
            enviarData(0, floatArrayOf())
            enqueRest(object : RestCallBack {
                override fun httpOK(vararg objects: Any?) {
                    val num = objects[0] as Int
                    val floats = objects[1] as FloatArray
                    val model = objects[2] as Model
                }

                override fun errorHttp(exception: Exception?, vararg objects: Any?) {
                    exception?.let { show(it.message) }
                }

                override fun successed(vararg objects: Any) {
                    show(objects[0].toString())
                }

                override fun failed(vararg objects: Any) {
                    show(objects[0].toString())
                }
            })
        }
    }

    fun MathCallback() {
        MyController.resolverProblema(3, 4, object : MathCallback() {
            override fun Mathsuccessed(vararg objects: Any) {
                show(objects[0].toString())
            }

            override fun Mathfailed(exception: NumberFormatException, vararg objects: Any?) {
                show(exception.message)
            }
        })
    }

    fun ObjectCallback() {
        MyController.resolverObject(null, object : ObjectCallback() {
            override fun successed(vararg objects: Any) {
                show(objects[0].toString())
            }

            override fun failed(vararg objects: Any) {
                show(objects[0].toString())
            }
        })
    }

    fun AbstractCallback() {
        MyController.resolverAbstract(3, "34", object : AbstractCallback() {
            override fun successed(vararg objects: Any) {
                show(objects[0].toString())
            }

            override fun failed(exception: Exception?, vararg objects: Any?) {
                exception?.let { show(it.message) }
            }
        })
    }

    fun MethodParse() {
        MyController.getParsedObject("", object : MethodCall() {
            override fun onFail(vararg objects: Any) {
                objects[0].toString()
            }

            override fun onParse(exception: ParseException, vararg objects: Any?) {
                show(exception.message)
            }

            override fun onSuccess(vararg objects: Any) {
                objects[0].toString()
            }
        })
    }

    fun OtherMethodParse() {
        MyController.getHttp("", object : HttpCall<Model?>() {
            override fun errorHTTP(codeError: String, exception: Exception, vararg objects: Any) {
                show(codeError + " - " + exception.message + " - " + objects[0].toString())
            }

            override fun onResponse(model: Model?, vararg objects: Any) {
                show(objects[0].toString())
            }

        })
    }

    fun twoSteps() {
        MyController.twoSteps(object : PreCallback {
            override fun successed(vararg objects: Any) {
                show(objects[0].toString())
            }

            override fun failed(vararg objects: Any) {
                show(objects[0].toString())
            }
        }, object : PosCallback {
            override fun successed(vararg objects: Any) {
                show(objects[0].toString())
            }

            override fun failed(vararg objects: Any) {
                show(objects[0].toString())
            }
        })
    }

    fun interfaces() {
        TemplateController<Model, Model2>()
                .methodStatickCall(object : TemplateController.TemplateCallback<Model?, Model2?>() {
                    override fun methodPoorReturnY(x: Model?): Model2? {
                        return Model2()
                    }

                    override fun methodSuperReturnA(b: Model2?): Model? {
                        return Model()
                    }

                })
    }

    fun returnCallback() {
        val value = 3
        MyController.returnTripleCallback(value) { it: Int, iTriple: Int ->
            show("The triple of $it is $iTriple")
        }
    }

    private fun show(string: String?) {
        Toast.makeText(this@MainActivity, string, Toast.LENGTH_SHORT).show()
    }
}